add Invoice, PaymentMethod and updateSubscription

yosoyubik opened the pull request:
Hi,

First, thanks for creating this. I've recently started learning Haskell and your repo has helped a lot in easing the learning curve :)

I'm working on a project that needs the endpoints that I have included in this PR so I was wondering if you'd be willing to include those here. I added some tests as well, and I'll be happy to update those or anything else that you think would fit better with the rest of your code or Haskell idioms.

Thanks!

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+
+data PaymentIntent = PaymentIntent
+  { piPaymentMethod :: Maybe T.Text,
+    piStatus :: Maybe T.Text
```
```suggestion
    piStatus :: T.Text
```

Should we also include the ID here?

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+    iCollectionMethod :: T.Text,
+    iCustomer :: CustomerId,
+    iHostedInvoiceURL :: Maybe T.Text,
+    iCreated :: Maybe Int,
```
```suggestion
    iCreated :: Maybe TimeStamp,
```

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+    iHostedInvoiceURL :: Maybe T.Text,
+    iCreated :: Maybe Int,
+    iAmountDue :: Maybe Int,
+    iAmountPaid :: Maybe Int,
```
Should we also add the currency?

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+    iAmountDue :: Maybe Int,
+    iAmountPaid :: Maybe Int,
+    iStatus :: T.Text,
+    iPaymentIntent :: Maybe PaymentIntent
```
```suggestion
    iPaymentIntent :: Either PaymentIntentId PaymentIntent
```

That way, we can support both the expanded and non expanded case

yosoyubik replied:
This looks good to me, but I'm having some trouble coming up with a way to parse a JSON into this. Any suggestions on what would be the best way to deal with it?

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+data Card = Card
+  { cExpMonth :: Int
+  , cExpYear :: Int
+  , cLast4 :: String
```
```suggestion
  , cLast4 :: T.Text
```

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +167,67 @@ data PriceCreateRecurring
+  }
+  deriving (Show, Eq)
+
+data PaymentMethodCreate = PaymentMethodCreate
```
This is focused on creating cards and Stripe recommends using Stripe.js for this setup to avoid needing to be PCI compliant. I'd suggest either removing it for now or extracting the card fields to a `CardCreate` data structure.

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -184,6 +254,7 @@ data Subscription
   = Subscription
   { sId :: SubscriptionId
   , sCancelAtPeriodEnd :: Bool
+  , sCollectionMethod :: Maybe T.Text
```
```suggestion
  , sCollectionMethod :: T.Text
```

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -285,7 +367,19 @@ instance ToForm CustomerCreate where
 
 instance ToForm CustomerUpdate where
-  toForm = genericToForm (formOptions 2)
+  toForm cu =
```
ooc: does the automatic deriving not work here? my assumption it only doesn't work for nested hashes in arrays. Either way, we can also leave it as is :)

agrafix commented on stripe-servant/src/Stripe/Api.hs:
```diff
@@ -40,10 +42,20 @@ type PriceApi
-  :<|> StripeAuth :> QueryParam "customer" CustomerId :> Get '[JSON] (StripeList Subscription)
+type InvoiceApi =
+  StripeAuth :> Capture ":invoice_id" InvoiceId :> Get '[JSON] Invoice
+    :<|> StripeAuth :> QueryParam "customer" CustomerId :> QueryParam "expand[]" T.Text :> Get '[JSON] (StripeList Invoice)
```
Would this work too?

```suggestion
    :<|> StripeAuth :> QueryParam "customer" CustomerId :> QueryParam "expand" [T.Text] :> Get '[JSON] (StripeList Invoice)
```

yosoyubik replied:
GHCID complains when I try with your change

agrafix commented on stripe-servant/src/Stripe/Api.hs:
```diff
@@ -40,10 +42,20 @@ type PriceApi
+
+type PaymentMethodApi =
+  StripeAuth :> QueryParam "customer" CustomerId :> QueryParam "type" T.Text :> Get '[JSON] (StripeList PaymentMethod)
+    :<|> StripeAuth :> Capture ":payment_method_id" PaymentMethodId :> "attach" :> QueryParam "customer" CustomerId :> Post '[JSON] PaymentMethod
```
I think something like this would be cleaner:

```suggestion
    :<|> StripeAuth :> Capture ":payment_method_id" PaymentMethodId :> "attach" :> ReqBody '[FormUrlEncoded] PaymentMethodAttach :> Post '[JSON] PaymentMethod
```

with 

```
data PaymentMethodAttach 
     = PaymentMethodAttach
     { pmaCustomerId :: CustomerId
     }
```

agrafix reviewed (changes requested):
This is awesome, thank you very much for the contribution!

I've made a few suggestions for improvements before we merge this, what do you think?

yosoyubik commented:
@agrafix I think I've addressed all your comments, let me know if I missed any.

PS: I ended up removing  PaymentMethodCreate/Attach

PPS: forget that, i still need to handle how to parse this: `iPaymentIntent :: Maybe (Either PaymentIntentId PaymentIntent)`

yosoyubik commented:
@agrafix, I ended up replacing the Either type for the Intent with a Sum type, so I could use Aeson's Alternative `(<|>)` instance.

For the tests (Invoices and Payment Method) I didn't rely  on the data generated by the tests itself (for example, the first Invoice created won't have a Payment Intent, so I couldn't check for the case where there's data present) so I assume that there are environment variables with a CustomerId and PaymentMethodId, pointing to a Customer with a PaymentMethod attach and Invoices with attempted payments.

Let me know what you think.

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +170,76 @@ data PriceCreateRecurring
+  deriving (Show, Eq)
+
+data PaymentIntentOrId
+  = Id PaymentIntentId
+  | Intent PaymentIntent
```
```suggestion
  = PiId PaymentIntentId
  | PiIntent PaymentIntent
```

To avoid collisions

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +170,76 @@ data PriceCreateRecurring
+    iCreated :: Maybe TimeStamp,
+    iAmountDue :: Maybe Int,
+    iAmountPaid :: Maybe Int,
+    iCurrency :: Maybe T.Text,
```
This should always be present

agrafix commented on stripe-servant/src/Stripe/Resources.hs:
```diff
@@ -158,6 +170,76 @@ data PriceCreateRecurring
+  deriving (Show, Eq, ToJSON, FromJSON, ToHttpApiData)
+
+newtype InvoiceSettings = InvoiceSettings
+  { isDefaultPaymentMethod :: Maybe T.Text
```
```suggestion
  { isDefaultPaymentMethod :: Maybe PaymentMethodId
```

agrafix reviewed (commented):
👍  neat this is great! left a few more suggestions, but am happy to merge after

yosoyubik commented:
Done! :)

(PS: let me know when you push to Hackage)
